Bluetooth: Mesh: Implement Bridge Config Client

HaavardRei opened the pull request:
Implements the Bridge Configuration Client model.
Adds opcodes for all Bridge messages to `foundation.h`. Adds client-side support for these messages.
Adds a new Kconfig option for the Bridge Configuration Client model to configure the waiting time for respones for acked messages.

HaavardRei committed:
Bluetooth: Mesh: Implement Bridge Config Client

Implements the Bridge Configuration Client model.
Adds opcodes for all Bridge messages to `foundation.h`. Adds client-side
support for these messages.
Adds a new Kconfig option for the Bridge Configuration Client model to
configure the waiting time for respones for acked messages.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei commented on subsys/bluetooth/mesh/brg_cfg_cli.c:
```diff
@@ -31,47 +151,170 @@ const struct bt_mesh_model_cb _bt_mesh_brg_cfg_cli_cb = {
+
+	bt_mesh_model_msg_init(&msg, OP_BRIDGING_TABLE_ADD);
+	net_buf_simple_add_u8(&msg, directions);
+	net_buf_simple_add_le24(&msg, (net_idx1 | (net_idx2 << 12)));
```
net_idx1/2 is of type uint16_t, but `net_buf_simple_add_le24` expects a uint32_t. Is there a problem with inputting the argument as it is here?

I did some local testing and it seems the uint16_t are "promoted" such that there is no overflow, but we may want to explicitly cast them to uint32_t before the shifting operation?

alxelax replied:
I did minor prototyping in online gcc. Seems it works. It will be nice to get confirmation from C standard. I know the width of variables are adjusted to max type in equation, but I'm not sure from which stage.

HaavardRei replied:
(Discussed elsewhere): According to C99 Integer promotion is performed in bitwise shift operations. local testing indicates that this should work.

alxelax reviewed (commented):
LGTM just a couple of questions

HaavardRei commented on include/zephyr/bluetooth/mesh/brg_cfg_cli.h:
```diff
@@ -300,6 +300,18 @@ int bt_mesh_brg_cfg_cli_bridging_table_get(uint16_t net_idx, uint16_t addr, uint
+ *
+ *  @param timeout The new transmission timeout.
+ */
+void bt_mesh_brg_cfg_cli_timeout_set(int32_t timeout);
```
It's possible, though this functionality is the same for other mesh clients other than in name.
Do you propose adding a check that it is a positive value?

HaavardRei commented on doc/connectivity/bluetooth/api/mesh/models.rst:
```diff
@@ -12,6 +12,8 @@ used by network administrators to configure and diagnose mesh nodes.
 .. toctree::
    :maxdepth: 1
 
+   brg_cfg_cli
```
They are displayed in the order specified here, i. e. alphabetical.

HaavardRei committed:
Bluetooth: Mesh: Fix subnet bridge table status type

Changes status type to be status codes defined in Mesh Protocol 4.3.14
as defined for the `BRIDGING_TABLE_STATUS` message.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei commented:
Found a bug in the bridging table status struct, the status type should be status codes from section 4.3.14 of MshPRT, not the enum. Added the commit fixing it here.

omkar3141 commented on subsys/bluetooth/mesh/brg_cfg_cli.c:
```diff
@@ -5,12 +5,158 @@
+	subnets_list.net_idx_filter.net_idx = (net_idx_filter >> 4) & BIT_MASK(12);
+	subnets_list.start_idx = net_buf_simple_pull_u8(buf);
+
+	if (buf->len) {
```
Should at least be "3" octets or more.

omkar3141 commented on subsys/bluetooth/mesh/brg_cfg_cli.c:
```diff
@@ -5,12 +5,158 @@
+	uint32_t net_key_indices;
+
+	table_list.status = net_buf_simple_pull_u8(buf);
+	net_key_indices = net_buf_simple_pull_le24(buf);
+	table_list.net_idx1 = net_key_indices & BIT_MASK(12);
+	table_list.net_idx2 = (net_key_indices >> 12) & BIT_MASK(12);
```
Better to use key_idx_unpack_pair() function from foundation.h.

omkar3141 commented on subsys/bluetooth/mesh/brg_cfg_cli.c:
```diff
@@ -5,12 +5,158 @@
+	table_list.net_idx2 = (net_key_indices >> 12) & BIT_MASK(12);
+	table_list.start_idx = net_buf_simple_pull_le16(buf);
+
+	if ((table_list.status == STATUS_SUCCESS) && buf->len) {
```
`buf->len` should be in multiples of `4` octets.

HaavardRei replied:
Multiples of 5, no? 
![image](https://github.com/user-attachments/assets/40f0b606-8691-4f8b-9783-36ab9a386e47)

omkar3141 commented on subsys/bluetooth/mesh/brg_cfg_cli.c:
```diff
@@ -31,47 +191,170 @@ const struct bt_mesh_model_cb _bt_mesh_brg_cfg_cli_cb = {
+	return bt_mesh_msg_ackd_send(cli->model, &ctx, &msg, !size ? NULL : &rsp_ctx);
 }
 
 int bt_mesh_brg_cfg_cli_bridging_table_add(uint16_t net_idx, uint16_t addr, uint8_t directions,
```
We should use `bridging_table_entry` struct here to reduce count of params. It is generally not efficient to pass more than 4 params to function call unless absolutely needed.

HaavardRei replied:
I agree.

omkar3141 commented on subsys/bluetooth/mesh/brg_cfg_cli.c:
```diff
@@ -31,47 +191,170 @@ const struct bt_mesh_model_cb _bt_mesh_brg_cfg_cli_cb = {
+	return bt_mesh_msg_ackd_send(cli->model, &ctx, &msg, !rsp ? NULL : &rsp_ctx);
 }
 
 int bt_mesh_brg_cfg_cli_bridging_table_remove(uint16_t net_idx, uint16_t addr, uint16_t net_idx1,
```
Same here.

HaavardRei replied:
The `bridging_table_entry` struct contains the `direction` field which is not used in removal. Do you still think we should change this?

alxelax commented on subsys/bluetooth/mesh/brg_cfg_cli.c:
```diff
@@ -5,12 +5,158 @@
+	struct bt_mesh_bridging_table_status *rsp;
+	uint32_t net_key_indices;
+
+	table_status.status = (enum bt_mesh_subnet_bridge_state)net_buf_simple_pull_u8(buf);
```
This is quite arguable. I guess we need to discuss it in IRL before doing such changes. I suggest leave it `as is`. Since this is a collaboration branch, we can change it later.

HaavardRei commented on doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:
```diff
@@ -0,0 +1,9 @@
+.. _bluetooth_mesh_models_brg_cfg_srv:
+
+Bridge Configuration Server
+###########################
```
I only created the doc page for the server so that the link from the client wouldn't complain. This should be done when implementing the server.

HaavardRei commented on subsys/bluetooth/mesh/brg_cfg_cli.c:
```diff
@@ -5,12 +5,158 @@
+
+	table_status.status = (enum bt_mesh_subnet_bridge_state)net_buf_simple_pull_u8(buf);
+	table_status.entry.directions = net_buf_simple_pull_u8(buf);
+	net_key_indices = net_buf_simple_pull_le24(buf);
+	table_status.entry.net_idx1 = net_key_indices & BIT_MASK(12);
+	table_status.entry.net_idx2 = (net_key_indices >> 12) & BIT_MASK(12);
```
Didn't know about it, good tip!

HaavardRei committed:
Bluetooth: Mesh: Change bridge table add params

Change the parameters of the bridge table add function to take an
entry struct rather than the fields of an entry.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>
